### PR TITLE
ChatGPT Plugin for Ergo Node API

### PR DESCRIPTION
**Pull request by aslesarenko**
In this PR:
- [x] A simplified OpenAPI  (`openapi-ai.yaml` file) added with a small subset of API endpoint specifications (see [article on hackernoon](https://hackernoon.com/bridging-ai-and-blockchain-developing-a-plugin-for-chatgpt)  for discussion of why a separate specification is necessary)
- [x] plugin manifest configuration is added to `.well-known/ai-plugin.json`
- [x] special Routes implemented which are requested by ChatGPT UI:
    - [`http://localhost:9053](http://localhost:9053)/.well-known/ai-plugin.json` - to get manifest
    - [`http://localhost:9053](http://localhost:9053)/openapi.yaml` - to get specification of API endpoints
- [x] Additional Routes implemented for new `get` endpoints which was added in openapi-ai.yaml
- [x] Additional response headers configured for ErgoHttpService (CorsHandler used by Ergo node)

### CODE REVIEW

File: src/main/resources/api/openapi.yaml [hunk 1]
```diff
@@ -2559,15 +2613,15 @@ paths:
...
+      summary: Get header ids at the given height
       operationId: getFullBlockAt
       tags:
         - blocks
       parameters:
         - in: path
           name: blockHeight
           required: true
-          description: Height of a wanted block
+          description: Height of a block to retrive header ids
```
Comment (kushti) [inline src/main/resources/api/openapi.yaml:2624]: retrive => retrieve

**Review by kushti — CHANGES_REQUESTED** [review]

File: src/main/resources/.well-known/ai-plugin.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  "auth": {
+    "type": "none"
+  },
+  "api": {
+    "type": "openapi",
+    "url": "http://localhost:9052/openapi.yaml",
+    "is_user_authenticated": false
+  },
+  "logo_url": "https://cryptologos.cc/logos/ergo-erg-logo.png",
+  "contact_email": "legal@example.com",
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/resources/.well-known/ai-plugin.json:16]: The [logo url](https://cryptologos.cc/logos/ergo-erg-logo.png) is valid.
And I changed the email to team@ergoplatform.org

File: src/main/resources/.well-known/ai-plugin.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  "auth": {
+    "type": "none"
+  },
+  "api": {
+    "type": "openapi",
+    "url": "http://localhost:9052/openapi.yaml",
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/resources/.well-known/ai-plugin.json:12]: fixed

File: src/main/resources/api/openapi.yaml [hunk 4]
```diff
@@ -13,6 +13,9 @@ info:
     url: https://raw.githubusercontent.com/ergoplatform/ergo/master/LICENSE
 
 servers:
+  - url: http://localhost:9052
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/resources/api/openapi.yaml:16]: fixed

File: src/main/resources/api/openapi.yaml [hunk 5]
```diff
@@ -362,31 +365,32 @@ components:
...
             $ref: '#/components/schemas/ErgoTransactionDataInput'
         outputs:
-          description: Outputs of the transaction
+          description: Outputs of the transaction, i.e. box candidates to be created by this transaction.
           type: array
           items:
             $ref: '#/components/schemas/ErgoTransactionOutput'
         size:
-          description: Size in bytes
+          description: Size of ErgoTransacton in bytes
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/resources/api/openapi.yaml:393]: fixed

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala [hunk 6]
```diff
@@ -147,14 +151,24 @@ case class BlockchainApiRoute(readersHolder: ActorRef, ergoSettings: ErgoSetting
...
+    } else {
       ApiResponse(getTxsByAddress(address, offset, limit))
     }
   }
 
+  private def getTxsByAddressR: Route = (post & pathPrefix("transaction" / "byAddress") & ergoAddress & paging) { (address, offset, limit) =>
+    validateAndGetTxsByAddress(address, offset, limit)
+  }
+
+  private def getTxsByAddressR_get: Route = (pathPrefix("transaction" / "byAddress") & get & addressPass & paging) { (address, offset, limit) =>
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala:168]: here `_get` part is somewhat special suffix, and `_` make it separate form the name. 
Please suggest the name, I'm ok with anything you suggest.

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala [hunk 7]
```diff
@@ -205,14 +219,24 @@ case class BlockchainApiRoute(readersHolder: ActorRef, ergoSettings: ErgoSetting
...
+    } else {
       ApiResponse(getBoxesByAddress(address, offset, limit))
     }
   }
 
+  private def getBoxesByAddressR: Route = (post & pathPrefix("box" / "byAddress") & ergoAddress & paging) { (address, offset, limit) =>
+    validateAndGetBoxesByAddress(address, offset, limit)
+  }
+
+  private def getBoxesByAddressR_get: Route = (pathPrefix("box" / "byAddress") & get & addressPass & paging) { (address, offset, limit) =>
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala:236]: please suggest a name

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala [hunk 8]
```diff
@@ -318,6 +357,13 @@ case class BlockchainApiRoute(readersHolder: ActorRef, ergoSettings: ErgoSetting
     ApiResponse(getAddressBalanceTotal(address))
   }
 
+  val addressPass: Directive1[ErgoAddress] = pathPrefix(Segment).flatMap(handleErgoAddress)
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala:360]: fixed

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala [hunk 9]
```diff
@@ -318,6 +357,13 @@ case class BlockchainApiRoute(readersHolder: ActorRef, ergoSettings: ErgoSetting
...
+
+  private def getAddressBalanceTotalR_v2: Route =
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala:362]: Fixed, changed to `_get` suffix to be consistent with other names. If you suggest another name, will change this as well. See my reply above.

File: src/main/resources/api/openapi.yaml (hunk 1 above)
Comment (aslesarenko) [reply src/main/resources/api/openapi.yaml:2624]: fixed

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala:236]: same as above

File: src/main/resources/api/openapi-ai.yaml [hunk 10]
```diff
@@ -0,0 +1,1480 @@
+openapi: "3.0.2"
+
+info:
+  version: "5.0.10"
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/resources/api/openapi-ai.yaml]: done

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala:168]: done

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala]: done

File: src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/http/api/BlockchainApiRoute.scala]: changed suffix to GetRoute

**Review by kushti — APPROVED** [review]

**Pull request merged**